pennylane_lightning_gpu/src/simulator/StateVectorCudaMPI.hpp

```diff
@@ -941,6 +955,364 @@ class StateVectorCudaMPI
+
+    /**
+     * @brief Convert a global CSR (Compress Sparse Row) format matrix into
+     * local blocks. This operation should be conducted on the rank 0.
+     *
+     * @tparam index_type Integer type used as indices of the sparse matrix.
+     * @param num_row_blocks Number of local blocks per global row.
+     * @param num_col_blocks Number of local blocks per global column.
+     * @param csrOffsets_ptr Pointer to the array of row offsets of the sparse
+     * matrix. Array of size csrOffsets_size.
+     * @param columns_ptr Pointer to the array of column indices of the sparse
+     * matrix. Array of size numNNZ
+     * @param values_ptr Pointer to the array of the non-zero elements
+     *
+     * @return auto A vector of vector of CSRMatrix.
+     */
+    template <class index_type>
+    auto splitCSRMatrix(const size_t &num_row_blocks,
+                        const size_t &num_col_blocks,
+                        const index_type *csrOffsets_ptr,
+                        const index_type *columns_ptr,
+                        const std::complex<Precision> *values_ptr)
+        -> std::vector<std::vector<CSRMatrix<Precision, index_type>>> {
+
+        std::vector<std::vector<CSRMatrix<Precision, index_type>>>
+            splitSparseMatrix(
+                num_row_blocks,
+                std::vector<CSRMatrix<Precision, index_type>>(num_col_blocks));
+
+        size_t num_rows = size_t{1} << this->getTotalNumQubits();
+        size_t row_block_size = size_t{1} << this->getNumLocalQubits();
+        size_t col_block_size = row_block_size;
+
+        for (size_t row = 0; row < num_rows; row++) {
+            for (size_t col_idx = static_cast<size_t>(csrOffsets_ptr[row]);
+                 col_idx < static_cast<size_t>(csrOffsets_ptr[row + 1]);
+                 col_idx++) {
+
+                size_t current_global_row = row;
+                size_t current_global_col = columns_ptr[col_idx];
+                std::complex<Precision> current_val = values_ptr[col_idx];
+
+                size_t block_row_id = current_global_row / row_block_size;
+                size_t block_col_id = current_global_col / col_block_size;
+
+                size_t local_row_id = current_global_row % row_block_size;
+                size_t local_col_id = current_global_col % col_block_size;
+
+                if (splitSparseMatrix[block_row_id][block_col_id]
+                        .csrOffsets.size() == 0) {
+                    splitSparseMatrix[block_row_id][block_col_id].csrOffsets =
+                        std::vector<index_type>(row_block_size + 1, 0);
+                }
+
+                splitSparseMatrix[block_row_id][block_col_id]
+                    .csrOffsets[local_row_id + 1]++;
+                splitSparseMatrix[block_row_id][block_col_id].columns.push_back(
+                    local_col_id);
+                splitSparseMatrix[block_row_id][block_col_id].values.push_back(
+                    current_val);
+            }
+        }
+
```

Comment:
I wonder if we should consider adding OpenMP pragmas here and earlier at a later date.

Reply:
yes, we can get performance improvement here by adding OpenMP pragmas. Have marked it here.